### DISCUSSION TITLE
Default Assembly Workbench Opinion

### DISCUSSION THREAD

**Discussion by sliptonic** [Ondsel-Development/FreeCAD#2] [Blog Topics]
FreeCAD has a rich eco-system of add-on workbenches.  These are often great tools that let FreeCAD serve niche areas of specialization.  Creating assemblies isn't a niche activity.  It's likely that many or most users will eventually need to design multiple parts that fit together and assembly is an important tool for that workflow. 

While FreeCAD is blessed with an abundance of assembly options through the addon manager, we believe that a default, pre-installed, assembly workbench is in the best interest of FreeCAD users.

We invite discussion here about how to evaluate the various assembly tools available and compare them.

**Comment by howie-j** [comment]
Not much response yet, so I'll get the sphere rolling :)

If i understand correctly, you want a discussion on **how** to evaluate the different workbenches, not necessarily a detailed pro/cons discussion on each of them?

One approach is to create a specification for an "ideal" assembly workbench, then compare each existing solution to find which approach makes the most sense; build upon an existing workbench and borrow ideas, functions and code from others, or just start from scratch and reuse what can be reused.

The specification should be very high level, like UX, workflow, constraints, functions like interactive dragging, BOM creation, top down modelling approaches and such.

My personal opinion is that none of the existing assembly workbenches is good enough to be included as the default, and all of them has different strengths and weaknesses.

**Comment by frankmcc** [comment]
I'm somewhere between A2Plus and Assembly3.  Both are fairly intuitive and work with both axis and planes.  A2Plus seems to be affected more by the topological naming problem than Assembly3.  Assembly4 is just a confusing mess, making use of additional origins and aligning parts on them.  If the topological naming problem did not exist A2Plus would be my go to.  The work flow is simple, intuitive and fast.

**Reply by aiksiongkoh** [reply]
In the world of CAD, commercial or open source, what features of assembly constraints are desirable and not? Coming from SpaceClaim, Tangent, Align, Orient of different combinations of geometric elements seems intuitive. 

http://help.spaceclaim.com/dsm/4.0/en/Content/Printable/SpaceClaimAssemblyConstraintsChart.pdf

**Comment by Marodut** [comment]
I believe the main difference is that in AS4 the placement and the movement of each of the bodies is scripted by the user, whilst in AS3 the result is "whatever the solver comes up with".

Relinquishing control to a tool is extremely valuable because you can use the computed results as a quality check for the design itself. [More on the matter](https://forum.freecad.org/viewtopic.php?t=76537)

On how to tackle this pickle, I believe AS4 funcionality can be replicated within AS3 enviroment, kinda like a "beginner" menu that expands to a fully featured toolset.

> My personal opinion is that none of the existing assembly workbenches is good enough to be included as the default, and all of them has different strengths and weaknesses. @howie-j 

I agree, AS3 is riddled  with quirks and considerations that are absolutely NON trivial

-  the solver needs to be babysitted
- quirky constraint types
- It relies on strict and cumbersome hierarchies
- ...

It can be made production ready and it can be stable, but it's a big commitment to ask.

I'm not very familiar with AS4 and AS2+, but they seem better starting blocks if this is decision is time sensitive.

**Comment by mlong799** [comment]
I am basically a noob related to programming.  However, I have worked with 9 CAD applications.  4 of which had solid modeling capability.  In my opinion I would like to see the following capability.
1)  Plane to Plane alignment. (plane needs to be defined with a perpendicular ray to provide direction.  If this occurs then it can be reversible)
      Usually the "Planes" should be the surface of the parts.  However, if the origin plane of the part can also be reference it adds substancial strength.
      Additional, functionality of the planes being offset is even stronger function.
2)  Axial alignment. (Axes needs to be created as a vector, if this happens then the direction can be reversible)
3)  Circular Feature to Feature alignment.  This allows the ability to work with the common feature of cylinders and holes in alignment.
      Circle Feature needs to be defined with a perpendicular center point ray to provide direction.  This allows ability to reverse.
      Additional, functionality of the circular features being offset is even stronger function.
4)  Angular Feature utilizing the Plane w/ direction or Circle w/ direction to have angles.
Multiple use of items 1-4 will allow for definition of almost every connection type.

If you wanted to go for the gold the ability to define tangent relationship between two surfaces
    - A planed would be defined with direction as described above
    - A simple curved surface (a single axis and radius) would be defined with a concave/convex surface to provide a direction.
    This would allow a tangent relationship to be defined between two curved surfaces or a single curved surface and a plane.

I think this functionality would work for almost all situations.

**Comment by magnificu** [comment]
The assembly module should be a container of parts.
Assembly inside a single part as multiple bodies (like AS2+) is counterintuitive and makes the management of sub-assemblies very difficult.

1) Planar to planar surface alignment
2) Rotation parts axial alignment, including cylindrical holes

The axis and the surfaces should be recognized automatically without constructing planes or axis manually
It can be done with a single button that offers before confirmation the parts to be flipped.

3) Multi instantiation of a part in circular and matrix pattern.

In my view easiness of use should have priority over design in context

In Sketcher all the coincidence buttons might be grouped and offer options in the sub-window

**Comment by magnificu** [comment]
Does anyone started working on a native assembly module?
The pseudo assembly add-ons (A2plus, Assembly 3, 4) are a useless joke in my view.

I think overall FreeCAD has a lot of potential and I hope the ultimate goal of the project is to grow the community and user base around it, but for this more is needed.

Unfortunately, as it stands today, I find it unnecessarily difficult even for simple operations like constraining a sketch. 
It is doable like is? Yes, but...
Why there are so many buttons like vertical/horizontal/distance constraint? A single button should be sufficient to do all this operations.
Why it is not possible to constrain a line to the edge of a body shape without creating external geometry manually? 
Why should all constraints only use point vertices? Why isn't it possible to apply a constrain by clicking on the lines directly?

Maybe is easier to program it this way, but for the user is very odd.

The focus should be on easiness of use and avoiding reinventing the wheel just for the sake of being different.

Get inspiration from serious CAD software like: CATIA v5,  Solid Works, Siemens NX, PTC Creo etc...
There were armies of people brainstorming on how to make them simple, intuitive and fast (minimum clicks)

Again why re-inventing the wheel?

e.g. The compass in CATIA v5 is a very powerful tool that can translate, rotate parts in assembly.

In FreeCAD the equivalent would be "The Cube", but can only select views... 
Maybe this will change with the new native assembly module :)

**Comment by Solar-Mann** [comment]
I have used most of the commercial CAD packages and stuck with SolidWorks for some years before switching entirely to FreeCAD (ASM3). I agree with most comments above, but at the end I have a different workflow and I'm using constraints rather to prepare animations and speed up the positioning of parts for file export. FreeCAD isn't perfect at all, but the others have plenty of shortcomings too which isn't a surprise since they all do essentially the same. 

However, FreeCAD "feels" as if there is a lack of coherence between the different workbenches and a lot of time is wasted by trying to remember how it was done last time. A2+ has a very intuitive workflow, compact and useful UI, but AMS3 remains the favorite for many reasons beside the assembly workbench.

Performance, stability and simplicity are on the wishlist. Importing A2+, A4 and A3 assemblies should be considered (if feasible).
